daemon/api.go

```diff
@@ -585,21 +585,50 @@ func deleteOp(c *Command, r *http.Request) Response {
 	}
 }
 
+type licT struct {
+	Intro   string
+	License string
+	Agreed  bool
+}
+
+func (*licT) Error() string {
+	// licT isn't an error, strictly speaking, but can
+	// be converted into one for convenience
+	return "license agreement required"
+}
+
 type packageInstruction struct {
+	progress.NullProgress
```

Comment:
we also have Notify on progress.Meter; in some places the interface is
called "interacter", fwiw; agreed the progress.Meter name isn't good.

We are not “doing a round trip inside snappy.Install just so it calls
Agreed on the progress meter, and we know from the outside whether
this will go wrong or not”; you need to read the manifest (and check
the system) to know whether you need to present a license agreement to
the user or not.

On 17 November 2015 at 12:43, Gustavo Niemeyer <notifications@github.com> wrote:
>
> In daemon/api.go:
>
> >  type packageInstruction struct {
> > + progress.NullProgress
>
> I see, thanks. So we do want to silence the progress metering aspect, but are using the progress.Meter interface to ask questions to the user. The are two things that makes this feel slightly uncomfortable:
>
> We have an Agreed function on progress.Meter, requiring interaction with the user (not progress metering, really)
>
> We are doing a round trip inside snappy.Install just so it calls Agreed on the progress meter, and we know from the outside whether this will go wrong or not. Perhaps we could have the license agreement logic completely out of snappy.Install.
>
> That said, it sounds fine to not do that now.
>
> —
> Reply to this email directly or view it on GitHub.